odbc/src/api/data.rs

```diff
@@ -0,0 +1,647 @@
+                second: dt.second() as u16,
+                fraction: (dt.nanosecond() as f32 * 0.000001) as u32,
+            };
+            set_output_fixed_data(&data, target_value_ptr, str_len_or_ind_ptr)
+        }
+        CDataType::Time | CDataType::TypeTime => {
+            let dt = data.to_date();
+            let data = Time {
+                hour: dt.hour() as u16,
+                minute: dt.minute() as u16,
+                second: dt.second() as u16,
+            };
+            set_output_fixed_data(&data, target_value_ptr, str_len_or_ind_ptr)
+        }
+        CDataType::Date | CDataType::TypeDate => {
+            let dt = data.to_date();
+            let data = Date {
+                year: dt.year() as i16,
+                month: dt.month() as u16,
+                day: dt.day() as u16,
+            };
+            set_output_fixed_data(&data, target_value_ptr, str_len_or_ind_ptr)
+        }
+        _ => {
+            mongo_handle.add_diag_info(ODBCError::Unimplemented("unimplemented data type"));
+            SqlReturn::ERROR
+        }
+    }
+}
+
+///
+/// input_text_to_string converts an input cstring to a rust String.
+/// It assumes nul termination if the supplied length is negative.
+///
+/// # Safety
+/// This converts raw C-pointers to rust Strings, which requires unsafe operations
+///
+/// for now this is cfg(test), when/if we decide to provide an ascii driver this will
+/// see use in actual code.
+#[allow(clippy::uninit_vec)]
+#[cfg(test)]
+pub unsafe fn input_text_to_string(text: *const Char, len: usize) -> String {
+    if (len as isize) < 0 {
+        let mut dst = Vec::new();
+        let mut itr = text;
+        {
+            while *itr != 0 {
+                dst.push(*itr);
+                itr = itr.offset(1);
+            }
+        }
+        return String::from_utf8_unchecked(dst);
+    }
+
+    let mut dst = Vec::with_capacity(len);
+    dst.set_len(len);
+    copy_nonoverlapping(text, dst.as_mut_ptr(), len);
+    String::from_utf8_unchecked(dst)
+}
+
+///
+/// input_wtext_to_string converts an input cstring to a rust String.
+/// It assumes nul termination if the supplied length is negative.
+///
+/// # Safety
+/// This converts raw C-pointers to rust Strings, which requires unsafe operations
+///
+#[allow(clippy::uninit_vec)]
+pub unsafe fn input_wtext_to_string(text: *const WChar, len: usize) -> String {
+    if (len as isize) < 0 {
+        let mut dst = Vec::new();
+        let mut itr = text;
+        {
+            while *itr != 0 {
+                dst.push(*itr);
+                itr = itr.offset(1);
+            }
+        }
+        return String::from_utf16_lossy(&dst);
+    }
+
+    let mut dst = Vec::with_capacity(len);
+    dst.set_len(len);
+    copy_nonoverlapping(text, dst.as_mut_ptr(), len);
+    String::from_utf16_lossy(&dst)
+}
+
+///
+/// set_sql_state writes the given sql state to the [`output_ptr`].
+///
+/// # Safety
+/// This writes to a raw C-pointer
+///
```

Comment:
If we ever do commit to an ascii driver, we'll need to change this to `set_sql_wstate` or some such.

Reply:
Or, actually, I guess these could be templated.